src/main/java/org/apache/commons/lang3/builder/Reflection.java

```diff
@@ -41,4 +90,44 @@ static Object getUnchecked(final Field field, final Object obj) {
         }
     }
 
+    /**
+     * Check whether the class internas of the given object can be accessed. This might return false in Java9
+     * and thus lead to a difference in behaviour if modularity is activated or not.
+     * But it's the best we can do.
+     *
+     * @return {@code true} if the given class internas not accessible
+     */
+    static boolean isNonIntrospectibleClass(Object o) {
+        return o != null && isNonIntrospectibleClass(o.getClass());
+    }
+
+    /**
+     * Check whether the given class internas can be accessed. This might return false in Java9
+     * and thus lead to a difference in behaviour if modularity is activated or not.
+     * But it's the best we can do.
+     *
+     * @return {@code true} if the given class internas not accessible
+     */
+    static boolean isNonIntrospectibleClass(Class<?> clazz) {
+        if (KNOWN_NON_INTROSPECTIBLE_CLASSES.contains(clazz)) {
+            return true;
+        }
+        Boolean isAccessible = nonIntrospectibleClasses.get(clazz);
+        if (isAccessible != null) {
+            return isAccessible;
+        }
+        try {
+            final Field[] declaredFields = clazz.getDeclaredFields();
+            for (Field f : declaredFields) {
+                f.setAccessible(true);
+            }
+
+            nonIntrospectibleClasses.put(clazz, false);
```

Comment:
This looks problematic since the map will grow uncontrollably, possibly until every class on your classpath gets referenced, there's no way to know really.

Reply:
this is a trade-off vs performance. it will only cache those classes which got hit in a reflection equals, toString, etc. Thus in practice it is a very limited set of classes I'd say. Can you think of a possible place where we can hold such a cache at least during a single invocation? Maybe inside the builder and handing the Map as another parameter to this method?